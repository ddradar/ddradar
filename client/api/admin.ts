import type { NuxtHTTPInstance } from '@nuxt/http'

import type { SongInfo } from '~/api/song'
import type { NotificationSchema } from '~/core/db/notification'

type NotificationRequest = Omit<
  NotificationSchema,
  'id' | 'sender' | 'timeStamp'
> & {
  id?: string
  timeStamp?: number
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postNotification
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: NotificationRequest
) {
  return $http.$post<NotificationRequest>('/api/v1/admin/notification', {
    ...body,
    sender: 'SYSTEM',
  })
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postSongInfo
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: SongInfo
) {
  return $http.$post<SongInfo>('/api/v1/admin/songs', body)
}
