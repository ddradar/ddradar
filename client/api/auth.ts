import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

type AuthResult = {
  clientPrincipal: Api.ClientPrincipal | null
}

/**
 * Call "Get Credentials" API.
 * @see https://docs.microsoft.com/azure/static-web-apps/user-information
 */
export async function getClientPrincipal(
  $http: Pick<NuxtHTTPInstance, '$get'>
) {
  const { clientPrincipal } = await $http.$get<AuthResult>('/.auth/me')
  return clientPrincipal
}
