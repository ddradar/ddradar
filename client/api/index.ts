export const apiPrefix = '/api/v1'
