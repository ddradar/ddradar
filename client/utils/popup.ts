import type { BuefyNamespace } from 'buefy/types'

const option = { position: 'is-top', hasIcon: true } as const

export function success(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-success', ...option })
}

export function warning(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-warning', ...option })
}

export function danger(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-danger', ...option })
}
