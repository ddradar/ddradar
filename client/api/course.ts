import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

export function getCourseType(type: number) {
  return type === 1 ? 'NONSTOP' : type === 2 ? '段位認定' : ''
}

/**
 * Call "Get Course List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses/
 */
export function getCourseList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  series?: 16 | 17,
  type?: 1 | 2
) {
  const searchParams = new URLSearchParams()
  if (series) searchParams.append('series', `${series}`)
  if (type) searchParams.append('type', `${type}`)
  return $http.$get<Api.CourseListData[]>(`${apiPrefix}/courses`, {
    searchParams,
  })
}

/**
 * Call "Get Course Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses__id/
 */
export function getCourseInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.CourseInfo>(`${apiPrefix}/courses/${id}`)
}
