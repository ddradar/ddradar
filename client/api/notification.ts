import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Notification List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification/
 */
export function getNotificationList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  topOnly?: boolean
) {
  const query = topOnly ? '?scope=top' : ''
  return $http.$get<Api.NotificationListData[]>(
    `${apiPrefix}/notification${query}`
  )
}

/**
 * Call "Get Notification Info" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification__id/
 */
export function getNotificationInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.NotificationInfo>(`${apiPrefix}/notification/${id}`)
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification--post/
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: Omit<Api.NotificationBody, 'sender'>
) {
  return $http.$post<Api.NotificationBody>(`${apiPrefix}/notification`, {
    ...body,
    sender: 'SYSTEM',
  })
}
