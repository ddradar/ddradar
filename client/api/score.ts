import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--get/
 */
export function getChartScore(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  scope?: 'private' | 'medium' | 'full'
) {
  const query = scope ? `?scope=${scope}` : ''
  return $http.$get<Api.ScoreInfo[]>(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}${query}`
  )
}

/**
 * Call "Post Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--post/
 */
export function postChartScore(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  score: Api.ScoreBody
) {
  return $http.$post(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}`,
    score
  )
}

/**
 * Call "Delete Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--delete/
 */
export function deleteChartScore(
  $http: Pick<NuxtHTTPInstance, 'delete'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4
) {
  return $http.delete(`/api/v1/scores/${songId}/${playStyle}/${difficulty}`)
}

/**
 * Call "Post Song Scores" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id--post/
 */
export function postSongScores(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  scores: Api.ScoreListBody[]
) {
  return $http.$post<Api.ScoreInfo[]>(`${apiPrefix}/scores/${songId}`, scores)
}
