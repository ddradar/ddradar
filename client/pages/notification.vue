<template>
  <section class="section">
    <h1 class="title">{{ $t('pageTitle') }}</h1>
    <section>
      <b-table
        :data="messages"
        paginated
        per-page="20"
        striped
        :loading="$fetchState.pending"
        detailed
        detail-key="id"
        show-detail-icon
      >
        <b-table-column v-slot="props" field="icon">
          <b-button :type="props.row.type" :icon-left="props.row.icon" />
        </b-table-column>
        <b-table-column v-slot="props" field="title" :label="$t('title')">
          {{ props.row.title }}
        </b-table-column>
        <b-table-column v-slot="props" field="timeStamp" :label="$t('date')">
          {{ props.row.date }}
        </b-table-column>
        <b-table-column
          v-slot="props"
          :visible="$accessor.isAdmin"
          label="Edit"
        >
          <nuxt-link :to="`/admin/notification/${props.row.id}`">
            <b-icon icon="pencil-box-outline" />
          </nuxt-link>
        </b-table-column>

        <template #detail="props">
          <!-- eslint-disable-next-line vue/no-v-html -->
          <div class="content" v-html="props.row.body" />
        </template>

        <template #empty>
          <section v-if="$fetchState.pending" class="section">
            <b-skeleton animated />
            <b-skeleton animated />
            <b-skeleton animated />
          </section>
          <section v-else class="section">
            <div class="content has-text-grey has-text-centered">
              <p>{{ $t('noData') }}</p>
            </div>
          </section>
        </template>
      </b-table>
    </section>
  </section>
</template>

<i18n>
{
  "ja": {
    "pageTitle": "お知らせ一覧",
    "title": "タイトル",
    "date": "投稿日時",
    "noData": "データがありません"
  },
  "en": {
    "pageTitle": "All Notification List",
    "title": "Title",
    "date": "Date",
    "noData": "No data"
  }
}
</i18n>

<script lang="ts">
import type { Api } from '@ddradar/core'
import marked from 'marked'
import { Component, Vue } from 'nuxt-property-decorator'
import type { MetaInfo } from 'vue-meta'

import { getNotificationList } from '~/api/notification'
import { unixTimeToString } from '~/utils/date'
import * as popup from '~/utils/popup'

type NotificationDetail = Omit<Api.NotificationListData, 'timeStamp'> & {
  date: string
}

@Component({ fetchOnServer: false })
export default class UserListPage extends Vue {
  messages: NotificationDetail[] = []

  /* istanbul ignore next */
  head(): MetaInfo {
    return {
      title: this.$t('pageTitle') as string,
    }
  }

  /** Call "Get Notification API" once. */
  async fetch() {
    try {
      const messages = await getNotificationList(this.$http)
      this.messages = messages.map(d => ({
        id: d.id,
        type: d.type,
        icon: d.icon,
        title: d.title,
        body: marked(d.body),
        date: unixTimeToString(d.timeStamp),
      }))
    } catch (error) {
      popup.danger(this.$buefy, error.message ?? error)
    }
  }
}
</script>
