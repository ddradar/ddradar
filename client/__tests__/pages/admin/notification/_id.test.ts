import { notification } from '@ddradar/core/__tests__/data'
import type { Context } from '@nuxt/types'
import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'

import { getNotificationInfo, postNotification } from '~/api/notification'
import EditorPage from '~/pages/admin/notification/_id.vue'
import * as popup from '~/utils/popup'

jest.mock('~/api/notification')
jest.mock('~/utils/popup')
const localVue = createLocalVue()
localVue.use(Buefy)

describe('pages/admin/notification/_id.vue', () => {
  describe('snapshot test', () => {
    test('renders correctly', () => {
      // Arrange
      const data = () => ({ ...notification })
      const wrapper = mount(EditorPage, { localVue, data })

      // Act - Assert
      expect(wrapper).toMatchSnapshot()
    })
  })

  // Computed
  describe('get pageTitle()', () => {
    test.each([
      ['', 'Add Notification'],
      ['foo', 'Update Notification'],
    ])('/%s returns "%s"', (id, expected) => {
      // Arrange
      const data = () => ({ id })
      const wrapper = shallowMount(EditorPage, { localVue, data })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.pageTitle).toBe(expected)
    })
  })
  describe('get hasError()', () => {
    test('({ title: "title", body: "body" }) returns false', () => {
      // Arrange
      const data = () => ({ title: 'title', body: 'body' })
      const wrapper = shallowMount(EditorPage, { localVue, data })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.hasError).toBe(false)
    })
    test.each([
      {},
      { title: 'title' },
      { body: 'body' },
      { title: 'title', body: '' },
    ])('(%p) returns true', d => {
      // Arrange
      const data = () => ({ ...d })
      const wrapper = shallowMount(EditorPage, { localVue, data })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.hasError).toBe(true)
    })
  })

  // LifeCycle
  describe('asyncData()', () => {
    beforeAll(() => mocked(getNotificationInfo).mockResolvedValue(notification))
    beforeEach(() => mocked(getNotificationInfo).mockClear())

    test('/ does not call getNotificationInfo()', async () => {
      // Arrange
      const wrapper = shallowMount(EditorPage, { localVue })
      const ctx = { params: {} } as Context

      // Act
      const result = await wrapper.vm.$options.asyncData!(ctx)

      // Assert
      expect(result).toBeUndefined()
      expect(mocked(getNotificationInfo)).not.toBeCalled()
    })
    test(`/${notification.id} calls getNotificationInfo()`, async () => {
      // Arrange
      const wrapper = shallowMount(EditorPage, { localVue })
      const ctx = { params: { id: notification.id } } as unknown as Context

      // Act
      const result = await wrapper.vm.$options.asyncData!(ctx)

      // Assert
      expect(result).toStrictEqual({
        id: notification.id,
        pinned: notification.pinned,
        type: notification.type,
        icon: notification.icon,
        title: notification.title,
        body: notification.body,
        timeStamp: notification.timeStamp,
      })
      expect(mocked(getNotificationInfo)).toBeCalled()
    })
  })

  // Method
  describe('saveNotification()', () => {
    const data = () => ({ ...notification })
    const mocks = { $buefy: {}, $http: {} }
    const wrapper = shallowMount(EditorPage, { localVue, data, mocks })
    beforeEach(() => {
      mocked(postNotification).mockClear()
      mocked(popup.success).mockClear()
      mocked(popup.danger).mockClear()
    })

    test('calls popup.success()', async () => {
      // Arrange
      mocked(postNotification).mockResolvedValue({ ...notification })

      // Act
      // @ts-ignore
      await wrapper.vm.saveNotification()

      expect(mocked(postNotification)).toBeCalled()
      expect(mocked(popup.success)).toBeCalledWith(mocks.$buefy, 'Success!')
      expect(mocked(popup.danger)).not.toBeCalled()
    })
    test('calls popup.danger()', async () => {
      // Arrange
      const error = 'Error'
      mocked(postNotification).mockRejectedValue(error)

      // Act
      // @ts-ignore
      await wrapper.vm.saveNotification()

      expect(mocked(postNotification)).toBeCalled()
      expect(mocked(popup.success)).not.toBeCalled()
      expect(mocked(popup.danger)).toBeCalledWith(mocks.$buefy, error)
    })
  })
})
