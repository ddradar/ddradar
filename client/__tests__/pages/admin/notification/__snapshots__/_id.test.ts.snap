// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/admin/notification/_id.vue snapshot test renders correctly 1`] = `
<section class="section">
  <h1 class="title">Update Notification</h1>
  <div class="field"><label class="label">Title</label>
    <div class="control is-clearfix"><input type="text" autocomplete="on" required="required" class="input">
      <!---->
      <!---->
      <!---->
    </div>
    <!---->
  </div>
  <div class="field"><label class="label">Body</label>
    <div class="control is-clearfix"><textarea required="required" class="textarea"></textarea>
      <!---->
      <!---->
      <!---->
    </div>
    <!---->
  </div>
  <div class="field">
    <!----><label class="switch is-rounded"><input type="checkbox" true-value="true" value="false"><span class="check"></span><span class="control-label">ピン留めする</span></label>
    <!---->
  </div>
  <div class="field"><label class="label">Type</label>
    <div class="control"><span class="select"><select><!----><option value="is-info">情報</option> <option value="is-warning">警告</option> <option value="is-danger">エラー</option></select></span>
      <!---->
    </div>
    <!---->
  </div>
  <div class="field"><label class="label">Icon</label>
    <div class="field">
      <!---->
      <div class="field-body">
        <div class="field is-grouped">
          <!---->
          <div class="control has-icons-right is-clearfix"><input type="text" autocomplete="on" class="input">
            <!----><span class="icon is-right"><i class="mdi mdi-info mdi-24px"></i></span>
            <!---->
          </div> <a type="button" href="https://materialdesignicons.com/" class="button is-text">
            <!----><span>
        参考ページ
      </span>
            <!---->
          </a>
          <!---->
        </div>
      </div>
      <!---->
    </div>
    <!---->
  </div>
  <div class="field">
    <!----><button type="button" class="button is-success">
      <!----><span>
      Save
    </span>
      <!---->
    </button>
    <!---->
  </div>
</section>
`;
